[Woo POS][Modularization] Remove all ServiceLocator usage from POS code

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

More context: pdfdoF-7It-p2
Continuation of https://github.com/woocommerce/woocommerce-ios/pull/15955

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

- Removing the remaining `ServiceLocator` usage and injecting dependencies to POS from the entry point. These 6 were enough to cover all the `ServiceLocator` usage:

```swift
public protocol POSDependencyProviding {
    var analytics: POSAnalyticsProviding { get }
    var stores: POSStoresProviding { get }
    var currency: CurrencySettings { get }
    var featureFlags: POSFeatureFlagProviding { get }
    var session: POSSessionManagerProviding { get }
    var connectivity: ConnectivityObserver { get }
}
```

### Payments

- I'm leaving `CardPaymentService` and related dependencies outside of the POS module. At least for now. They would require a separate refactoring to untangle from IPP, and at the moment, they are injected through `CardPresentPaymentFacade` dependency that is defined for `POS`.

<img width="485" height="197" alt="image" src="https://github.com/user-attachments/assets/a56c6750-aab4-4de3-94c5-827ee3c7ec61" />

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Review a general approach. Thorough testing is not necessary since it's not targeting the trunk.

- CI should succeed
- POS should launch
- POS events should be tracked

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad 18.5 simulator, iPad Air M2 26 device

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Create POSServiceAdapter to inject all services through environment

staskus committed:
Access analytics through serviceAdapter

staskus committed:
Prepare environment keys for used Woo app Services

staskus committed:
Remove ServiceLocator.currencySettings from PointOfSaleOrderController

staskus committed:
Remove ServiceLocator.currencySettings from PointOfSaleCollectCashView

staskus committed:
Remove ServiceLocator.currencySettings from CollectCashViewHelper

staskus committed:
Remove ServiceLocator.currencySettings from CardPresentPaymentCollectOrderPaymentUseCaseAdaptor

staskus committed:
Implement POSStoresProviding

It only requires a single dispatch method that is used within POS application

staskus committed:
Remove unused StoresManager from PointOfSaleOrderController

staskus committed:
Implement POSPreviewStores

staskus committed:
Move POS payment services that depend on IPP to POS/Adaptor folder

CardPresentPaymentService and its dependencies have a direct relationship with the current IPP implementation which would need to be refactored separately to be able to move to the POS module. For now, keeping it within the Woo app.

staskus committed:
Rename adapters to adaptors for consistency

staskus committed:
Remove ServiceLocator.featureFlagService dependency from POS

staskus committed:
Remove store from POS dependencies

staskus committed:
Remove ServiceLocator.stores.sessionManager dependency from POS

staskus committed:
Move ConnectivityObserver to WooFoundation for reusability and set to environment in POS

staskus committed:
Move DefaultConnectivityObserverTests to WooFoundationTests

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Clarify POSAnalyticsKey documentation for SwiftUI

staskus committed:
Use dependency injection for POSServiceLocatorAdaptor to enable testing

staskus committed:
Document analytics parameter in observer init for GameControllerBarcodeObserver

staskus commented:
It's yesterday's work. It's an open question whether it's worth reviewing now or waiting for the end of the week, if the end result is fine, and then going back and reviewing the PRs. I'll try to stack a couple on top of each other. cc @jaclync @iamgabrielma 


----

In case you're interested,

Today, I progressed more and reached a point where moving the majority of the POS code resulted in successfully building PointOfSale module. I did it by moving all POS files to PointOfSale module, trying to build, and adding any dependencies that are missing to the module or commenting stuff out until everything successfully built. I found it hard to detect all the dependencies from the Woo target within the POS code, so this way the compiler helped me.

However, this is all untidy. Woo app dependencies within POS will either need to be:
- Moved to a shared module (WooFoundation, Yosemite...)
- Injected to POS through the adaptor or some other way
- Duplicated in the POS target
- ... or just move away outside the POS folder since they are not really used

I make a decision on a case-by-case basis for these dependencies, which really depends on whether it's just an extension or a large class with a lot of entangled dependencies.

Here's a list of remaining dependencies that I'm in process of addressing:
* DeepLinkNavigator
* SafariView
* SupportForm
* CardPresentPaymentsOnboardingViewModel
* CardPresentPaymentError
* CardPresentPaymentInvalidatablePaymentOrchestrator
* CardPresentPaymentsConnectionControllerManager
* POSCollectOrderPaymentAnalytics
* CollectOrderPaymentAnalytics
* WooAnalytics event constants between IPP and POS
* OrderTotalsCalculator
* PaymentCaptureCelebrationProtocol
* VersionHelpers
* POSEntryPointEligibilityCheckerProtocol
* SelectedSiteSettingsProtocol
* WooAnalytics+WooApp
* View+Conditionals
* .product(name: "CocoaLumberjackSwift", package: "CocoaLumberjack")
* scrollVerticallyIfNeeded
* ScrollableVStack
* KeyboardObserver
* POSCollectOrderPaymentPreviewAnalytics
* POSPreviewEntryPointEligibilityChecker
* PointOfSaleEntryPointView
* UIImage+Woo
* CollectOrderPaymentUseCaseError
* WCSettingsWebView
* String+Helpers
* Array+Helpers
* View+Measurements
* AddEditCoupon view
* POSCouponCreationSheetModifier
* BottomSheetListSelector
* AddEditCouponViewMode
* POSProductFactory
* WooRoundedBorderTextFieldStyle
* SwiftUI Shimmer
* View+RoundedBorder
* readableCountry extension CountryCode
* ReceiptEligibilityUseCase
* IndefiniteCircularProgressViewStyle
* View+autofocustextmodifier
* ProductImageThumbnail
* FormattableAmountTextFieldViewModel
* FormattableAmountTextField
* ServiceLocator currency settings
* PriceFieldFormatter
* CurrencySettings
* Sanitized
* WaitingTimeTracker
* trackSize(size.) modifier
* WooConstants


Once all these dependencies find their place to be used within the PointOfSale module, we could have a v1. A PointOfSale module with the majority of POS-only code, which is dependent on other modules within the app, and has some injected dependencies from the Woo app.

jaclync commented on Modules/Sources/PointOfSale/DependencyProviders/POSDependencyProviding.swift:
```diff
@@ -23,36 +25,15 @@ public protocol POSAnalyticsProviding {
 public protocol POSStoresProviding {
-    var sessionManager: POSSessionManagerProviding { get }
-    // Add other stores manager methods as needed
+    func dispatch(_ action: Yosemite.Action)
```
was hoping not having to have this in the POS module, but I guess there's still some usage 😅

staskus replied:
You're actually right. I think we can avoid it, I think it's only been used in the card payment code, which will be outside the POS now. 👍

jaclync commented on Modules/Sources/PointOfSale/Environment/POSEnvironmentKeys.swift:
```diff
@@ -25,9 +29,79 @@ private struct DefaultPOSAnalytics: POSAnalyticsProviding {
+}
+
+// Default implementations for testing/previews
+private struct DefaultPOSStores: POSStoresProviding {
```
nit: since we also have `POSPreviewServices` that includes individual dependencies for previews, how about removing the default dependencies here if the purpose is for SwiftUI previews? Previews that need it can initialize `POSPreviewServices` and set the required environment keys. If we miss setting an environment key, having the app crash would make it more discoverable than some features silently failing IMO.

staskus replied:
I don't know if we can avoid setting a default value, but I will check. What we could do is to reuse these Default stores within the Preview service. 

>  If we miss setting an environment key, having the app crash would make it more discoverable than some features silently failing IMO.

That's true, but it's also unlikely that we forget, given that we inject it at the entry point. When we add a new dependency when developing a feature, we will see that something is not working as it should.

jaclync commented on WooCommerce/Classes/POS/Adaptors/POSServiceLocatorAdaptor.swift:
```diff
@@ -0,0 +1,117 @@
+    init(featureFlagService: FeatureFlagService = ServiceLocator.featureFlagService,
+         storesManager: StoresManager = ServiceLocator.stores,
+         connectivityObserver: ConnectivityObserver = ServiceLocator.connectivityObserver,
+         currencySettings: CurrencySettings = ServiceLocator.currencySettings,
```
just a note that `ServiceLocator.currencySettings` could change if the currency settings (site settings) sync is triggered throughout the store life cycle. The POS eligibility check should ensure the currency is eligible for POS, but if the check is within the POS module, we might need to update the currency in `POSDependencyProviding` in case the merchant changes the currency settings in core and retry in the app to be eligible for POS. We can test it when we move the eligibility check code.

staskus replied:
I need to check how it works. We could always return the latest `ServiceLocator.currencySettings` when users need currency settings, instead of injecting the value during initialization. I'll check if that helps. I didn't understand if the issues you described below are related to my changes or happen in general even on trunk.

jaclync commented on WooCommerce/Classes/POS/TabBar/POSTabCoordinator.swift:
```diff
@@ -93,11 +93,12 @@ private extension POSTabCoordinator {
                                                                             stores: storesManager,
-                                                                            collectOrderPaymentAnalyticsTracker: collectOrderPaymentAnalyticsTracker)
+                                                                            collectOrderPaymentAnalyticsTracker: collectOrderPaymentAnalyticsTracker,
+                                                                            currencySettings: serviceAdaptor.currency)
```
similar to a previous comment, the currency could change if the merchant updates it to make the store eligible for POS. Let's make sure to test the unsupported currency flow later. If there's an issue, we can move the currency parameter from the initializer to `CardPresentPaymentFacade.collectPayment` since that seems to be where the currency is used for formatting order amount.

jaclync replied:
I tested the unsupported currency flow, and the payment failed with SCPError 9020 (Stripe API error). Payment succeeded in the next POS launch after the currency was updated to be eligible for POS. We can fix this separately.


https://github.com/user-attachments/assets/32121ccd-74d1-41c7-a283-1c7e4844222b

jaclync replied:
Another issue with payments in a happy path, iPad Pro 11in M4 iOS 18.5 simulator with simulated reader enabled: after payment success (heard a cha-ching and payment success event logged), the UI is stuck on the processing screen:


https://github.com/user-attachments/assets/54c12662-4e27-4eb4-9020-e054860e37e5

staskus replied:
🤔 Thanks, I'll check what could be the issue.

staskus replied:
> I tested the unsupported currency flow, and the payment failed with SCPError 9020 (Stripe API error). Payment succeeded in the next POS launch after the currency was updated to be eligible for POS. We can fix this separately.

I'll fix this by reverting the. Changes in the payment adaptor and card payment service. Those changes are no longer needed since this code will remain in the POS module.

This is a hard thing when trying to do refactoring in small chunks, each step may invalidate the previous one since it was only needed or it looked to be needed as an intermediate step.

> Another issue with payments in a happy path, iPad Pro 11in M4 iOS 18.5 simulator with simulated reader enabled: after payment success (heard a cha-ching and payment success event logged), the UI is stuck on the processing screen:

I'm getting this on trunk as well.

jaclync reviewed (approved):
Glad seeing ServiceLocator usage removed from POS! Just an issue from the unsupported currency flow that can be fixed later.

> It's an open question whether it's worth reviewing now or waiting for the end of the week, if the end result is fine, and then going back and reviewing the PRs. I'll try to stack a couple on top of each other.

Having incremental work reviewed along the way sounds fine to me, the PR size has been manageable so far given a bunch of diffs are just moving files and updating dependencies.

> Here's a list of remaining dependencies that I'm in process of addressing:

Let me/Gabriel (if he has time) know if we can help pick up some dependencies in the list, especially for the ones we worked on. I can pick up the eligibility related ones if you haven't started. For shared views, we can also consider creating a new module like WooFoundationUI/WooUI though maybe keeping all in existing modules is more practical.

staskus commented:
@jaclync, thank you very much again for review and catching some broken cases, really valuable!

> Let me/Gabriel (if he has time) know if we can help pick up some dependencies in the list, especially for the ones we worked on. 

Okay. I will try to leave a task with a list after the day of what could be tackled next. I think it would be counter-productive to work on these things simultaneously due to many conflicts. However, we could tackle it one after another in different time zones.

staskus committed:
Remove POSStoresProviding from POS dependencies

staskus committed:
Revert removal of ServiceLocator in CardPresentPaymentCollectOrderPaymentUseCaseAdaptor

staskus committed:
Replace previews with default empty values

staskus committed:
Structure POSServiceLocatorAdaptor to always use ServiceLocator when services are accessed

staskus commented:
@jaclync  Thanks again for an incredible review.

- Removed stores. dispatch from POS dependencies
- Addressed currency issue by continuing to use ServiceLocator within card service and adapters since it won't be moved to POS. Additionally, change the way the adaptors are implemented to always make sure the latest service dependency is returned
- Addressed non-succeeding card payments here https://github.com/woocommerce/woocommerce-ios/pull/15967
